src/analysis/bits-bounds-transfer-function.h

```diff
@@ -0,0 +1,367 @@
+#ifndef wasm_analysis_bits_bounds_transfer_function_h
+#define wasm_analysis_bits_bounds_transfer_function_h
+
+#include <unordered_map>
+
+#include "bits-bounds-lattice.h"
+#include "ir/bits.h"
+#include "ir/stack-utils.h"
+#include "stack-lattice.h"
+#include "visitor-transfer-function.h"
+
+namespace wasm::analysis {
+
+class MaxBitsTransferFunction
+  : public VisitorTransferFunc<MaxBitsTransferFunction,
+                               StackLattice<MaxBitsLattice>,
+                               AnalysisDirection::Forward> {
+  MaxBitsLattice& bitsLattice;
+
+public:
+  std::unordered_map<Expression*, Index> exprMaxBounds;
+
+  MaxBitsTransferFunction(MaxBitsLattice& bitsLattice)
+    : bitsLattice(bitsLattice) {}
+
+  void visitConst(Const* curr) {
+    MaxBitsLattice::Element currElement = bitsLattice.getBottom();
+
+    bool addInformation = true;
+    switch (curr->type.getBasic()) {
+      case Type::i32:
+        currElement.setUpperBound(
+          32 - curr->value.countLeadingZeroes().geti32(), curr->value);
+        break;
+      case Type::i64:
+        currElement.setUpperBound(
+          64 - curr->value.countLeadingZeroes().geti64(), curr->value);
+        break;
+      default: {
+        addInformation = false;
+      }
+    }
+
+    if (collectingResults && addInformation) {
+      exprMaxBounds[curr] = currElement.upperBound;
+    }
+
+    currState->push(std::move(currElement));
+  }
+
+  void visitBinary(Binary* curr) {
+    MaxBitsLattice::Element right = currState->pop();
+    MaxBitsLattice::Element left = currState->pop();
+
+    MaxBitsLattice::Element currElement = bitsLattice.getBottom();
+
+    bool addInformation = true;
+    switch (curr->op) {
+      case RotLInt32:
+      case RotRInt32:
+      case SubInt32: {
+        currElement.setUpperBound(32);
```

Comment:
Looks like there's room for improvement here if the right side is known to be 0. Is that worth improving now, or are we trying to match 1:1 with the existing getMaxBits as much as possible?

Reply:
Added a TODO.